### PR TITLE
analyzer: Add a package curation provider based on SW360

### LINKED ISSUE #2210: Provide a curation provider for SW360
In order to receive curations from SW360, a curation provider for SW360 has to be implemented. As SW360 rest client the library to be build in https://github.com/eclipse/antenna/issues/423 should be used.

Acceptance criteria:
- [ ] For identified components, known metadata stored in a SW360 instance can be provided as curations
- [ ] The SW360 instance to use can be configured including authentication information
- [ ] Priority of SW360 information is higher then Clearly Defined data because SW360 is the internal knowledge base for component metadata

### PR DESCRIPTION
**Pull request by bs-ondem**
This PR adds a package curation provider based on SW360. In order to use it set the flag `--sw360-curations` and add a config file with `-c <path-to-ort.conf>` in order to pass SW360 connection configurations like URLs, credentials and proxy. 

Resolves #2210.

Signed-off-by: Onur Demirci <onur.demirci@bosch.io>

### CODE REVIEW

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import org.ossreviewtoolkit.model.Identifier
+import org.ossreviewtoolkit.model.PackageCuration
+import org.ossreviewtoolkit.model.PackageCurationData
+import org.ossreviewtoolkit.model.RemoteArtifact
+import org.ossreviewtoolkit.model.Hash
+import org.ossreviewtoolkit.model.HashAlgorithm
+import org.ossreviewtoolkit.model.config.Sw360Configuration
+
+
+class Sw360PackageCurationProvider(sw360Configuration: Sw360Configuration?) : PackageCurationProvider {
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:40]: This class needs to be documented, see the other package curation providers for examples.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                        vcs = null,
+                        comment = "Provided by SW360."
+                    )
+                )
+            )
+        } ?: emptyList()
+    }
+
+    private fun getSourceArtifactFromRelease(release: SW360SparseRelease, downloadPath: File) : RemoteArtifact? =
+        sw360ReleaseClient.getReleaseById(release.releaseId).takeIf { it.isPresent }?.let { it ->
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:83]: Remove `it ->` or replace it with a meaningful name.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            val fullRelease = it.get()
+            val attachments = fullRelease.embedded?.attachments
+            attachments?.filter { it.attachmentType == SW360AttachmentType.SOURCE }
+                ?.map { attachment ->
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:87]: Use `forEach` instead of `map`, we don't need to create a list here.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                    sw360ReleaseClient.downloadAttachment(fullRelease, attachment, downloadPath.toPath())
+                        .takeIf { it.isPresent }.let {
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:89]: Should be `?.let`.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                                url = it.toString(),
+                                hash = Hash(
+                                    value = attachment.sha1,
+                                    algorithm = HashAlgorithm.SHA1
+                                )
+                            )
+                        }
+                }
+            RemoteArtifact.EMPTY
+        }
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:100]: Having a `return` inside an expression function is a bit unusual and long expression functions can become hard to understand. I believe making this not an expression function and using `if` instead of `takeIf` could make this function easier to read.

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt [hunk 6]
```diff
@@ -100,6 +104,11 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
         help = "Whether to fall back to package curation data from the ClearlyDefine service or not."
     ).flag()
 
+    private val useSw360Curations by option(
+        "--sw360-curations",
+        help = "Whether to fall to package curation data from the SW360 service or not."
```
Comment (mnonnenmacher) [inline cli/src/main/kotlin/commands/AnalyzerCommand.kt:109]: "fall **back** to"

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt [hunk 7]
```diff
@@ -19,19 +20,26 @@
...
+import com.fasterxml.jackson.annotation.JsonAlias
+
 data class AnalyzerConfiguration(
     /**
      * If set to true, ignore the versions of used command line tools. Note that this might lead to erroneous
      * results if the tools have changed in usage or behavior. If set to false, check the versions to match the
      * expected versions and fail the analysis on a mismatch.
      */
-    val ignoreToolVersions: Boolean,
+    val ignoreToolVersions: Boolean = false,
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/AnalyzerConfiguration.kt:31]: If you set the default value you should add "Defaults to `false`" to the docs of the property.
This and the change below should go into a separate commit.

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt [hunk 8]
```diff
@@ -19,19 +20,26 @@
...
 
     /**
      * Enable the analysis of projects that use version ranges to declare their dependencies. If set to true,
      * dependencies of exactly the same project might change with another scan done at a later time if any of the
      * (transitive) dependencies are declared using version ranges and a new version of such a dependency was
      * published in the meantime. If set to false, analysis of projects that use version ranges will fail.
      */
-    val allowDynamicVersions: Boolean
+    val allowDynamicVersions: Boolean = false,
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/AnalyzerConfiguration.kt:39]: If you set the default value you should add "Defaults to `false`" to the docs of the property.

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt [hunk 9]
```diff
@@ -19,19 +20,26 @@
...
+
+    val sw360Configuration: Sw360Configuration? = null,
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/AnalyzerConfiguration.kt:41]: Please add docs for the property.

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt [hunk 10]
```diff
@@ -19,19 +20,26 @@
...
+
+    @JsonAlias("analyzer")
+    val options: Map<String, Map<String, String>>? = null
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/AnalyzerConfiguration.kt:44]: Do you use this property anywhere?

File: model/src/main/kotlin/config/OrtConfiguration.kt [hunk 11]
```diff
@@ -29,6 +30,10 @@ import java.io.File
  * The configuration model for all ORT components.
  */
 data class OrtConfiguration(
+    /**
+     * The configuration of the scanner.
+     */
+    val analyzer: AnalyzerConfiguration? = null,
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/OrtConfiguration.kt:36]: I like that you make the analyzer configurable in the config file, but this should happen in a separate commit.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+        sw360Configuration?.password,
+        sw360Configuration?.clientId,
+        sw360Configuration?.clientPassword,
+        ServiceFactory().createHttpClient(
+            sw360Configuration?.proxyHost.isNullOrEmpty() && sw360Configuration?.proxyPort!! > 0,
+            sw360Configuration?.proxyHost,
+            sw360Configuration?.proxyPort!!
+        ),
+        ServiceFactory.getObjectMapper()
+    )
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:55]: I expect that if the `sw360Configuration` is `null` this provider will not work. I would suggest to add an init block like this to the class to provide the user with meaningful feedback in this case:
```
init {
  requireNotNull(sw360Configuration) {
    "SW360 is not configured. To use curations from SW360 configure it in the ORT configuration file."
  }
}
```

File: model/src/main/kotlin/config/Sw360Configuration.kt [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.model.config
+
+data class Sw360Configuration(
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/Sw360Configuration.kt:22]: Please add documentation for this class. All public API should be documented.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt (hunk 1 above)
Comment (bs-ondem) [reply analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:39]: Do you mean documentation using KDoc or in the `docs` module?

Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:42]: I meant KDoc.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                )
+            )
+        } ?: emptyList()
+    }
+
+    private fun getSourceArtifactFromRelease(release: SW360SparseRelease, downloadPath: File): RemoteArtifact? {
+        val releaseById = sw360ReleaseClient.getReleaseById(release.releaseId)
+        if (releaseById.isPresent) {
+            val fullRelease = releaseById.get()
+            val attachments = fullRelease.embedded?.attachments ?: emptySet()
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:93]: Use `.orEmpty()` instead of `?: emptySet()` for consistency with other ORT code.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+            if (attachments.isNotEmpty()) {
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:94]: This check is unnecessary.

**Review by mnonnenmacher — COMMENTED** [review]
Sorry for the late review @bs-ondem , I have missed that you updated the PR. I have resolved all comments which you have addressed, but a few are remaining.

**Comment by bs-ondem** [comment]
Never mind @mnonnenmacher i was also out of office. :) 

I added your last comments in this PR and separated some changes into new commits.

File: model/src/main/kotlin/config/OrtConfiguration.kt (hunk 11 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/config/OrtConfiguration.kt:36]: The `reference.conf` needs to be updated to include the analyzer configuration. For reference, and to have at least a basic test for it in `OrtConfigurationTest`. Also, the values should be used in the analyzer. To correctly implement that I think the `--allow-dynamic-versions` and `--ignore-tool-versions` command line parameters of the analyzer should be removed, because they can now also be set using `-P`. @sschuberth Any opinion on that?

File: model/src/main/kotlin/config/OrtConfiguration.kt [hunk 16]
```diff
@@ -29,6 +30,10 @@ import java.io.File
  * The configuration model for all ORT components.
  */
 data class OrtConfiguration(
+    /**
+     * The configuration of the scanner.
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/OrtConfiguration.kt:34]: Replace "scanner" with "analyzer".

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt [hunk 17]
```diff
@@ -19,19 +20,22 @@
 
 package org.ossreviewtoolkit.model.config
 
+import com.fasterxml.jackson.annotation.JsonAlias
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/AnalyzerConfiguration.kt]: Commit message:

> In order to make the AnalyzerConfiguration deserialisable, default values are set.

The reason is not that the class is generally not deserializable without the default values, but that you want to be able to omit some of the values in the config file, especially with the SW360 configuration you are adding in the next commit.

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt [hunk 18]
```diff
@@ -144,13 +151,21 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
...
         val analyzer = Analyzer(analyzerConfig)
 
         val curationProvider = FallbackPackageCurationProvider(
             listOfNotNull(
                 packageCurationsFile.takeIf { it.isFile }?.let { FilePackageCurationProvider(it) },
-                ClearlyDefinedPackageCurationProvider().takeIf { useClearlyDefinedCurations }
+                ClearlyDefinedPackageCurationProvider().takeIf { useClearlyDefinedCurations },
+                config.analyzer?.sw360Configuration?.let {
+                    Sw360PackageCurationProvider(config.analyzer?.sw360Configuration).takeIf { useSw360Curations }
+                }
```
Comment (mnonnenmacher) [inline cli/src/main/kotlin/commands/AnalyzerCommand.kt:168]: The order in this list defines the priority of the providers, I would assume that the SW360 provider should be above the ClearlyDefined provider.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

**Comment by bs-ondem** [comment]
After the discussion with @sschuberth following changes will be done:
- To ensure that the data from SW360 is really curated, the `ClearingState` of the release must be checked. It should only use the data from SW360 if the `ClearingState` is `APPROVED`.
- Take the homepage URL from the component of the release. 
- The sources for the package which are downloaded as attachments from SW360 should not be saved in a temporary directory in order to use these as a file URL for the `RemoteArtifact`, but the `RemoteArtifact` should take directly the URL of the attachment given from the release. 
- Also take the binary attachment from SW360 to create the `RemoteArtifact` for `binaryArtifact`.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+class Sw360PackageCurationProvider(sw360Configuration: Sw360StorageConfiguration?) : PackageCurationProvider {
+
+    init {
+        requireNotNull(sw360Configuration) {
+            "SW360 is not configured. To use curations from SW360, configure it in the ORT configuration file."
+        }
+    }
+
+    private val sw360JsonMapper = jsonMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+    private val sw360ConnectionFactory = createSw360Connection(sw360Configuration!!, sw360JsonMapper)
```
Comment (oheger-bosch) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:55]: The field name is misleading; it is a SW360Connection, not a ConnectionFactory.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+    private val sw360ComponentClient = sw360ConnectionFactory.componentAdapter
+    private val sw360ReleaseClient = sw360ConnectionFactory.releaseAdapter
```
Comment (oheger-bosch) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:57]: It is probably a matter of taste, but I prefer not having too many member fields. The sw360JsonMapper is certainly unnecessary. With regards to the clients, you can consider storing only the connection and access them from there when needed.

**Review by oheger-bosch — COMMENTED** [review]
Two minor points; otherwise looks good to me.

**Comment by sschuberth** [comment]
@bs-ondem please rebase this on latest master to get rid of some test failures.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+/**
+ * A [PackageCurationProvider] for curated package meta-data from the configured SW360 instance using the REST API.
+ */
+class Sw360PackageCurationProvider(sw360Configuration: Sw360StorageConfiguration?) : PackageCurationProvider {
+
+    init {
+        requireNotNull(sw360Configuration) {
+            "SW360 is not configured. To use curations from SW360, configure it in the ORT configuration file."
+        }
+    }
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:52]: How about making the `sw360Configuration` argument non-nullable to get rid of this code?

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+                    PackageCuration(
+                        id = pkgId,
+                        data = PackageCurationData(
+                            declaredLicenses = sw360Release
+                                .embedded
+                                ?.licenses
+                                ?.filterNotNull()
+                                ?.map { license -> license.shortName }
+                                ?.toSortedSet() ?: sortedSetOf(),
+                            homepageUrl = getHomepageOfRelease(sw360Release) ?: "",
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:77]: Please use `.orEmpty()`.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+                listOf(
+                    PackageCuration(
+                        id = pkgId,
+                        data = PackageCurationData(
+                            declaredLicenses = sw360Release
+                                .embedded
+                                ?.licenses
+                                ?.filterNotNull()
+                                ?.map { license -> license.shortName }
+                                ?.toSortedSet() ?: sortedSetOf(),
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:76]: How about

    declaredLicenses = sw360Release.embedded?.licenses
        ?.mapNotNullTo(sortedSetOf()) { it?.shortName },

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+            ?.let { it ->
+                val attachment = it[0]
+                return RemoteArtifact(
+                    url = attachment.links.self.href,
+                    hash = Hash(
+                        value = attachment.sha1,
+                        algorithm = HashAlgorithm.SHA1
+                    )
+                )
+            }
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:107]: How about

        release.embedded?.attachments.orEmpty().singleOrNull { it.attachmentType == type }?.let { attachment ->
            return RemoteArtifact(
                url = attachment.links.self.href,
                hash = Hash(
                    value = attachment.sha1,
                    algorithm = HashAlgorithm.SHA1
                )
            )
        }

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+
+    private fun getHomepageOfRelease(release: SW360Release): String? =
+        sw360Connection.componentAdapter.getComponentById(release.componentId)
+            ?.get()
+            ?.homepage
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:112]: Should be `.orElse(null)?.homepage`.

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt [hunk 26]
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (C) 2017-2019 HERE Europe B.V.
+ * Copyright (C) 2020 Bosch.IO GmbH
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/AnalyzerCommand.kt:3]: Nit: "-2021".

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt [hunk 27]
```diff
@@ -153,12 +160,20 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
...
+            allowDynamicVersions,
+            config.analyzer?.sw360Configuration
+        )
         val analyzer = Analyzer(analyzerConfig)
 
         val curationProvider = FallbackPackageCurationProvider(
             listOfNotNull(
                 packageCurationsFile.takeIf { it.isFile }?.let { FilePackageCurationProvider(it) },
+                config.analyzer?.sw360Configuration?.let {
+                    Sw360PackageCurationProvider(config.analyzer?.sw360Configuration).takeIf { useSw360Curations }
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/AnalyzerCommand.kt:175]: Just use `it` instead of `config.analyzer?.sw360Configuration` here.

File: detekt-rules/src/main/kotlin/OrtPackageNaming.kt [hunk 28]
```diff
@@ -63,6 +63,7 @@ class OrtPackageNaming : Rule() {
             "helper-cli" -> ".helper"
             "nexus-iq" -> ".nexusiq"
             "spdx-utils" -> ".spdx"
+            "sw360" -> ".sw360"
```
Comment (sschuberth) [inline detekt-rules/src/main/kotlin/OrtPackageNaming.kt:66]: Is this needed? This should be the default.

File: helper-cli/build.gradle.kts [hunk 29]
```diff
@@ -52,6 +53,16 @@ repositories {
             includeGroup("org.gradle")
         }
     }
+
```
Comment (sschuberth) [inline helper-cli/build.gradle.kts:56]: Reminder: Probably not necessary anymore when rebasing onto `master` after https://github.com/oss-review-toolkit/ort/pull/3235 got merged.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Comment by sschuberth** [comment]
@bs-ondem, https://github.com/oss-review-toolkit/ort/pull/3235 was merged, so please rebase this once more onto `master`.

File: analyzer/build.gradle.kts [hunk 30]
```diff
@@ -1,6 +1,7 @@
 /*
  * Copyright (C) 2017-2019 HERE Europe B.V.
  * Copyright (C) 2019 Bosch Software Innovations GmbH
+ * Copyright (C) 2020 Bosch.IO GmbH
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply analyzer/build.gradle.kts:4]: Does not seem to be addressed.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 31]
```diff
@@ -0,0 +1,113 @@
...
+                            comment = "Provided by SW360."
+                        )
+                    )
+                )
+            }
+            .orElse(emptyList())
+    }
+
+    private fun getAttachmentAsRemoteArtifact(release: SW360Release, type: SW360AttachmentType): RemoteArtifact? =
+        release.embedded?.attachments.orEmpty().singleOrNull { it.attachmentType == type }?.let { attachment ->
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:81]: It just occurred to me that we could simply replace `.orEmpty()` with `?` here.

File: helper-cli/build.gradle.kts [hunk 32]
```diff
@@ -1,6 +1,7 @@
 /*
  * Copyright (C) 2017-2019 HERE Europe B.V.
  * Copyright (C) 2019 Bosch Software Innovations GmbH
+ * Copyright (C) 2020 Bosch.IO GmbH
```
Comment (sschuberth) [inline helper-cli/build.gradle.kts:4]: This change can be dropped now.

File: model/src/main/kotlin/config/OrtConfiguration.kt (hunk 11 above)
Comment (sschuberth) [reply model/src/main/kotlin/config/OrtConfiguration.kt:42]: Missing blank line after this line.

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt (hunk 17 above)
Comment (sschuberth) [reply model/src/main/kotlin/config/AnalyzerConfiguration.kt]: This import should not be added in this commit.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/build.gradle.kts [hunk 33]
```diff
@@ -1,6 +1,7 @@
 /*
  * Copyright (C) 2017-2019 HERE Europe B.V.
  * Copyright (C) 2019 Bosch Software Innovations GmbH
+ * Copyright (C) 2020-2021 Bosch.IO GmbH
```
Comment (sschuberth) [inline analyzer/build.gradle.kts:4]: Commit message nit: Drop the "." after "SW360" in the title.

File: model/src/test/assets/reference.conf [hunk 34]
```diff
@@ -12,6 +12,14 @@ ort {
   analyzer {
     ignoreToolVersions = true
     allowDynamicVersions = true
+    sw360Configuration {
```
Comment (sschuberth) [inline model/src/test/assets/reference.conf:15]: Nit: Missing blank line before this one.

File: model/src/test/kotlin/config/OrtConfigurationTest.kt [hunk 35]
```diff
@@ -48,6 +48,14 @@ class OrtConfigurationTest : WordSpec({
             ortConfig.analyzer shouldNotBeNull {
                 ignoreToolVersions shouldBe true
                 allowDynamicVersions shouldBe true
+                sw360Configuration shouldNotBeNull {
```
Comment (sschuberth) [inline model/src/test/kotlin/config/OrtConfigurationTest.kt:51]: Nit: Missing blank line before this one.

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt [hunk 36]
```diff
@@ -162,8 +151,8 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
 
         val config = globalOptionsForSubcommands.config
         val analyzerConfig = AnalyzerConfiguration(
-            ignoreToolVersions,
-            allowDynamicVersions,
+            config.analyzer?.ignoreToolVersions == true,
+            config.analyzer?.allowDynamicVersions == true,
             config.analyzer?.sw360Configuration
         )
         val analyzer = Analyzer(analyzerConfig)
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/AnalyzerCommand.kt]: I believe this can now be simplified to just

    val analyzer = Analyzer(config.analyzer ?: AnalyzerConfiguration())

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt [hunk 37]
```diff
@@ -36,4 +36,9 @@ data class AnalyzerConfiguration(
      * false.
      */
     val allowDynamicVersions: Boolean = false,
+
+    /**
+     * Configuration of the SW360 package curation provider.
+     */
+    val sw360Configuration: Sw360StorageConfiguration? = null
```
Comment (sschuberth) [inline model/src/main/kotlin/config/AnalyzerConfiguration.kt:43]: Due to this, this commit also needs to add `sw360_configuration: null` to the `config` section of expcted ORT result files to make tests pass. This should be the list of files:

```
$ git grep -F "    ignore_tool_versions:" -- "*.yml"
analyzer/src/funTest/assets/projects/external/directories-expected-output.yml:    ignore_tool_versions: false
analyzer/src/funTest/assets/projects/external/git-repo-expected-output.yml:    ignore_tool_versions: false
analyzer/src/funTest/assets/projects/external/sbt-multi-project-example-expected-output.yml:    ignore_tool_versions: false
analyzer/src/funTest/assets/projects/synthetic/gradle-all-dependencies-expected-result-with-curations.yml:    ignore_tool_versions: false
analyzer/src/funTest/assets/projects/synthetic/gradle-all-dependencies-expected-result.yml:    ignore_tool_versions: false
cli/src/funTest/assets/semver4j-analyzer-result.yml:    ignore_tool_versions: false
reporter/src/funTest/assets/gradle-all-dependencies-result.yml:    ignore_tool_versions: false
reporter/src/funTest/assets/static-html-reporter-test-input.yml:    ignore_tool_versions: false
scanner/src/funTest/assets/analyzer-result.yml:    ignore_tool_versions: false
scanner/src/funTest/assets/file-counter-expected-output-for-analyzer-result.yml:    ignore_tool_versions: false
```

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Comment by sschuberth** [comment]
> I believe we should always bulk fetch all in one go already now, because otherwise IMO it is instantly tech debt.

I agree single-fetching instead of bulk / batch requests is suboptimal. But we mainly need this for workflow demonstration purposes right now, so performance is not an issue. We'll either improve this later if it becomes a performance issue, or eventually remove it again before it really becomes tech debt.

> So far there could be the hidden assumption in the codebase that calling the provider twice with same parameters results in the same result always.

I'm not aware of such an assumption, and it would be wrong already before this PR, e.g. for the `ClearlyDefinedPackageCurationProvider`. So if something would need to get fixed, it would be these assumptions (if any), and not the providers, IMO.

File: analyzer/src/funTest/assets/projects/external/directories-expected-output.yml [hunk 38]
```diff
@@ -25,6 +25,7 @@ analyzer:
   config:
     ignore_tool_versions: false
     allow_dynamic_versions: false
+    sw360_configuration: null
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/assets/projects/external/directories-expected-output.yml:28]: We do not want credentials to be serialized into the ORT result files. As this change adds the SW360 config to the ORT result it should also add `@JsonProperty(access = JsonProperty.Access.WRITE_ONLY)` to the `password` and `clientPassword` properties of `Sw360StorageConfiguration`, similar to `PostgresStorageConfiguration.password`. There should also be a test for this to prevent that it accidentally breaks and credentials are added to the ORT results. See e.g. `HttpFileStorageConfigurationTest`. I believe there was a similar test for `PostgresStorageConfiguration` but I couldn't find it anymore, maybe it was removed in a refactoring.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 39]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.analyzer.curation
+
+import com.fasterxml.jackson.databind.DeserializationFeature
+import com.fasterxml.jackson.databind.ObjectMapper
+import org.eclipse.sw360.antenna.http.HttpClientFactoryImpl
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:24]: Empty line above.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt [hunk 40]
```diff
@@ -101,22 +101,11 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
         .convert { it.absoluteFile.normalize() }
         .configurationGroup()
 
-    private val allowDynamicVersions by option(
```
Comment (mnonnenmacher) [inline cli/src/main/kotlin/commands/AnalyzerCommand.kt:102]: Commit message: Typo in "dynammic"

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt (hunk 37 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/config/AnalyzerConfiguration.kt:43]: I think it should rather not serialize this property if it is null instead.

Comment (sschuberth) [reply model/src/main/kotlin/config/AnalyzerConfiguration.kt:43]: Agreed, that's probably better. Sorry @bs-ondem for me not coming up with that idea in the first place.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt (hunk 39 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:24]: BTW, this should now be checked automatically thanks to https://github.com/oss-review-toolkit/ort/pull/3494.

File: model/src/test/kotlin/config/Sw360StorageConfigurationTest.kt [hunk 41]
```diff
@@ -0,0 +1,47 @@
...
+        )
+        val yaml = yamlMapper.writeValueAsString(sw360StorageConfiguration).trim()
+
+        yaml shouldBe """
+                --- !<.Sw360StorageConfiguration>
+                rest_url: "url"
+                auth_url: "url"
+                username: "username"
+                client_id: "clientId"
+            """.trimIndent()
```
Comment (sschuberth) [inline model/src/test/kotlin/config/Sw360StorageConfigurationTest.kt:45]: I'd prefer a slight more generic check like `yaml.toLowerCase() shouldNotContain "password"`  here.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by sschuberth — APPROVED** [review]
LGTM, @mnonnenmacher do you also want to do a re-review?

**Review by mnonnenmacher — APPROVED** [review]
